File: src/main/scala/io/github/fherbreteau/gatling/ftp/FtpDsl.scala
package io.github.fherbreteau.gatling.ftp

import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.session.Expression
import io.github.fherbreteau.gatling.ftp.protocol.{FtpProtocol, FtpProtocolBuilder}

class FtpDsl {

  def ftp(implicit configuration: GatlingConfiguration): FtpProtocolBuilder = FtpProtocolBuilder(configuration)

  def ftp(requestName: Expression[String]) = Ftp(requestName)

  implicit def ftpProtocolBuilder2SftpProtocol(builder: FtpProtocolBuilder): FtpProtocol = builder.build

}

File: src/main/java/io/github/fherbreteau/gatling/ftp/javaapi/FtpDsl.java
package io.github.fherbreteau.gatling.ftp.javaapi;

import io.gatling.javaapi.core.Session;
import io.gatling.javaapi.core.internal.Expressions;
import io.github.fherbreteau.gatling.ftp.javaapi.protocol.FtpProtocolBuilder;

import jakarta.annotation.Nonnull;
import java.util.function.Function;

/**
 * The entrypoint of the Gatling FTP DSL
 */
public class FtpDsl {

    /**
     * Bootstrap a FTP protocol configuration
     */
    public static final FtpProtocolBuilder ftp =
            new FtpProtocolBuilder(io.github.fherbreteau.gatling.ftp.protocol.FtpProtocolBuilder.apply(
                    io.gatling.core.Predef.configuration()));

    private FtpDsl() {
    }

    /**
     * Bootstrap an FTP command configuration
     *
     * @param name the FTP command name, expressed as a Gatling Expression Language String
     * @return the next DSL step
     */
    @Nonnull
    public static Ftp ftp(@Nonnull String name) {
        return new Ftp(Expressions.toStringExpression(name));
    }

    /**
     * Bootstrap an FTP command configuration
     *
     * @param name the FTP command name, expressed as a function
     * @return the next DSL step
     */
    @Nonnull
    public static Ftp ftp(@Nonnull Function<Session, String> name) {
        return new Ftp(Expressions.javaFunctionToExpression(name));
    }

}
